[feat] engine: implementation of seekr for news, images and videos

Bnyro opened the pull request:
## What does this PR do?
* implement seekr as engine for news, videos and images
* uses the official seekr api
* the videos support is very basic due to api limitations - not sure if we can do anything about it
* seekr has two different api versions: v1 (used for images and videos) and v2 (used for news)

## Why is this change important?
* I don't have much experience with seekr, but the results seem to be decent

## How to test this PR locally?
* !senews World
* !seimg World
* !sevid World

## Related issues
closes #1644

Bnyro committed:
[feat] engine: implementation of seekr for news, images and videos

return42 commented on searx/engines/seekr.py:
```diff
@@ -0,0 +1,104 @@
+about = {
+    "website": 'https://seekr.com/',
+    "official_api_documentation": None,
+    "use_official_api": True,
```
When the API is official but not documented, from where did you get your know how of the details from the API .. reversed engendered?

Bnyro replied:
Not sure if the API is official, I didn't search deeper about it - I reverse engineered the API requests in the browser dev tools.
From my research now the API is probably not official, I just thought so because we need to provide an `apiKey`, which is always the same when using it in the browser however.

return42 replied:
> I reverse engineered the API requests in the browser dev tools.

This is what I also often do :) .. TBH I haven't thought in deep, but I think the intention the flag `use_official_api` is to show if the engine uses an official API that is documented (and versioned) .. on the opposite we have engines where we know there is an  API but we didn't use the API:

https://github.com/searxng/searxng/blob/c76c999a1eeecae265573b12e81ae74f56988820/searx/engines/google.py#L42-L43

Even the google engine is reversed engineered, it can help to understand the arguments of the API when we follow (more or less) the official API.

I will change the value of the flag in my review ..

return42 commented on searx/engines/seekr.py:
```diff
@@ -0,0 +1,104 @@
+# v2/newssearch, v1/imagetab, v1/videotab
+seekr_path = "newssearch"
+seekr_api_version = "v2"
+api_key = "srh1-22fb-sekr"
```
I think we still have too little experience with the API to offer official settings for it today.  If we release such settings today, then we also have to maintain them in the future and in the meantime some maintainers may have made their own (local) settings, which would then overwrite our bugfixes again.

If it is OK for you I would like to hard-code these values (as part of my review).

----

I am not sure yet, but maybe we can get the api-key via a fetch method too

Bnyro replied:
Sure, that makes sense!
However, I don't think we can fetch an API key somehow - it's still the same one as when the seekr issue got opened here 1.5 years ago, so I think it's fine to use that one.

allendema commented on searx/engines/seekr.py:
```diff
@@ -0,0 +1,104 @@
+def _news_response(json):
+    results = []
+
+    for result in json['expertResponses'][0]['advice']['categorySearchResult']['searchResult']['results']:
```
This will give out an error if nothing is found.

Maybe use something like this:

```py
results = []

if json['expertResponses'][0]['responseStatus'] == "NoResultsMatched":
       return results
```
...

return42 replied:
Hi @allendema thanks for pointing out / I just in review with this PR and fixed this issue already .. give me a few days for the review and I will come back here with a revision of the implementation .. and then it would be great if you test the revision (which will include major changes).

return42 committed:
[mod] improve seekr engines and add documentation

Tis patch adds some more fields to the result items and changed paging to the
``nextResultSet`` given in seekr's JSON response.

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

Bnyro commented on searx/engines/seekr.py:
```diff
@@ -53,52 +123,97 @@ def _images_response(json):
-    for result in json['expertResponses'][0]['advice']['categorySearchResult']['searchResult']['results']:
+    for result in search_results['results']:
+
+        gettext("Language")
```
That `gettext` seems to be unused?

return42 replied:
Oops .. I dropped the line :facepalm:
